proposals/infra/INF-0004-binding-resource-binding-diagnostics.md

```diff
@@ -0,0 +1,212 @@
+that obviously are not, or cannot contain, a valid resource type (types 
+like int, float, float4, etc), the only valid binding prefix is 'b', which
+adheres to legacy behavior. All other binding prefixes applied to such
+a type will emit an error diagnostic that "the given type '%0' cannot be
+bound as a resource". If the binding prefix is 'b', a warning will be 
+emitted instead, that will be treated as an error by default:
+"binding prefix 'b' used for resource type '%0', which cannot be used
+as a resource". Below are some examples:
+
+| non-intangible type | Binding statements | Diagnostic |
+|-|-|-|
+| `float f` | register(t0) | "error: 'float' is an invalid resource type for binding prefix 't'." |
+| `float f` | register(b0) | "warning: binding prefix 'b' used for resource type 'float', which cannot be used as a resource" |
+
+
+
+## Detailed design
+
+All the compiler has to work with is a Decl object. From the Decl object, 
+the ResourceAttr attribute can be fetched, with which the resource 
+class can be determined. Once this is known, we then check if the resource
+type is a user-defined type or if it is a standard resource type. If the 
+resource type is among the standard resource types (e.g., cbuffer, tbuffer,
+RWBuffer, etc.), then the binding prefix is checked for validity. The resource
+is also checked for sampleability (i.e., is it a texture resource?). If there is
+an incompatibility between resource class and binding prefix, a diagnostic will
+be emitted. Otherwise, the resource is a user-defined type, and so the type needs
+to be validated to see whether it can properly be assigned to the geiven binding 
+prefix. If the UDT does not have at least one resource that matches the given binding 
+prefixes, then an error will be emitted. The only remaining case is when there
+is no resource attribute, which implies the type is non-intangible. In this case,
+if the type is strictly numeric, the only acceptable binding prefix is 'b', which 
+is legacy behavior. In this case, a warning will be emitted that will be treated as 
+an error by default, which states that a numeric type is not usable as a resource. 
+Any other binding prefix will cause an error to be emitted, stating that the given 
+type cannot be bound as a resource.
+Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules
+flag. When this flag is active, and legacy behavior is present, a warning will 
+be emitted instead of an error.
+
+Below is a flowchart that describes the process that is used to determine
+what kind of diagnostic to emit in each case:
+
+```mermaid
+flowchart TD
+A[Type T bound to 
+binding prefix p] --> B{Does T have or 
+contain any type
+with the ResourceAttr
+attribute?}
+B -- Yes -->C{Is T a UDT?}
+B -- No --> M{Is T non-intangible?}
+C -- Yes -->D{Does T contain
+at least one 
+valid resource 
+for p?}
+C -- No -->G{"Is T sampleable 
+(is it a Texture* resource?)"}

```

Comment:
> is it a Texture* resource?

Presumably this has to come from the ResourceAttr?  This isn't doing a string compare against the typename is it?

Reply:
I am envisioning that a function will be made, "IsSampleable", that checks to see whether a given resource type is sampleable (I don't think such a function exists yet.) As of right now, it just so happens that a resource type is sampleable if and only if the resource kind name starts with "Texture". However, the value of resource kinds are just integers since they are defined in an enum. So, I am intending to implement the "IsSampleable" function as a comparison against the integer value of the resource kind component of the ResourceAttr attribute. The integer would be recovered from the ResourceAttr attribute, specifically the ResouceKind component.